Update README.md (#57)
m-tolstrup/buzzy: README.md

# buzzy

> This project was developed as a part of a master thesis at Aalborg University. The project focused on fuzzing eBPF technolgoies. For this, we targeted the PREVAIL verifier and uBPF virtual machine, in order to run a setup similar to a differential fuzzer, to find conformance issues between the two. The main bulk of code written for the project consisted of a eBPF program generator, and a parser to produce ELF files.

Reviewer: ```suggestion
This project was developed as a part of a master thesis at Aalborg University. The project focused on fuzzing eBPF technologies.
We targeted the PREVAIL verifier and uBPF virtual machine, to find conformance issues between the two. The main bulk of code written for the project consists of an eBPF program generator, and a parser to produce ELF files.
```

We will make the project report available when we have defended the thesis.

## Project Structure

> The buzzy project has the following folder structure:

Reviewer: ```suggestion
The Buzzy project has the following folder structure:
```

- **PREVAIL** (Submodule) (Polynomial-Runtime eBPF Verifier using an Abstract Interpretation Layer)
  - We use the PREVAIL verifier, to verify generated eBPF programs.
- **uBPF** (Submodule) (User space eBPF verifier and JIT-compiler for C programs)
  - We use the uBPF virtual machine to run the verified programs.
- **rbpf** (Submodule) (Rust eBPF)
  - We use the eBPF structures provided by this Rust eBPF crate to generate eBPF programs.
- **buzzy**

>   - Our fuzzing harness utilizing a generator and ELF file parser.

Reviewer: ```suggestion
  - Our fuzzing harness which utilizes a generator and ELF file parser.
```

  - `cargo +nightly fuzz run test/random/stack_sequence/random_maps`

## Trophies

> buzzy has found the following (confirmed) bugs:

Reviewer: ```suggestion
Buzzy has found the following (confirmed) bugs:
```

- [Inconsistency in load instruction handling between PREVAIL and uBPF](https://github.com/vbpf/ebpf-verifier/issues/484)
- [Segmentation fault (core dumped) - Null value in modulo operations](https://github.com/vbpf/ebpf-verifier/issues/493)

> ## Future Work

Reviewer: maybe remove, or not, its fine

We see the following areas as next the additions or changes for buzzy:

- **Move away from cargo-fuzz**: cargo-fuzz only ended up being used for what is essentially a looping mechanism. Implementing a version of this that would better suit buzzy would be greatly beneficial for the project.
- **Guided fuzzing**: Extending above, this setup could allow for better guided fuzzing.

> - **Extendned fault detection**: For the RPEVAIL/uBPF setup, buzzy only observed for conformance issues between the targets. No run time behavior or similar is captured.

Reviewer: ```suggestion
- **Extended fault detection**: For the PREVAIL/uBPF setup, Buzzy only observed conformance issues between the targets. No run time behavior or similar is captured.
```
